pkg/socket/client.go

```diff
@@ -0,0 +1,85 @@
+package socket
+
+import (
+	"bufio"
+	"crypto/tls"
+	"errors"
+	"net"
+	"time"
+)
+
+// New returns a new pointer to a socket client given the socket
+// type (IP, TCP, UDP, UNIX), a network address (IP/domain:port),
+// a timeout and a TLS config. It supports both IPv4 and IPv6 address
+// and reuses connection where possible.
+func New(config Config) *Socket {
+	return &Socket{
+		Config: config,
+		conn:   nil,
+	}
+}
+
+// Socket is the implementation of a socket client.
+type Socket struct {
+	Config
+	conn net.Conn
+}
+
+// Connect connects to the Socket address on the named network.
+// If the address is a domain name it will also perform the DNS resolution.
+// Address like :80 will attempt to connect to the localhost.
+// The config timeout and TLS config will be used.
+func (s *Socket) Connect() (err error) {
+	if s.TLSConf == nil {
+		s.conn, err = net.DialTimeout(string(s.Network), s.Address, s.Timeout)
+	} else {
+		var d net.Dialer
+		d.Timeout = s.Timeout
+		s.conn, err = tls.DialWithDialer(&d, string(s.Network), s.Address, s.TLSConf)
+	}
+	return err
+}
+
+// Disconnect closes the connection.
+// Any in-flight commands will be cancelled and return errors.
+func (s *Socket) Disconnect() (err error) {
+	if s.conn != nil {
+		err = s.conn.Close()
+		s.conn = nil
+	}
+	return err
+}
+
+// Command writes the command string to the connection and passed the
+// response bytes line by line to the process function. It uses the
+// timeout value from the Socket config and returns read, write and
+// timeout errors if any. If a timeout occurs during the processing
+// of the responses this function will stop processing and return a
+// timeout error.
+func (s *Socket) Command(command string, process Processor) error {
+	if err := write(command, s.conn, s.Timeout); err != nil {
+		return err
+	}
+	return read(s.conn, process, s.Timeout)
+}
```

Comment:
How do we handle closed connections? Or situations when we need to recreate the connection internally in order to write/send?

Also, this function panics if we don't call Connect(). If that is expected, I think we need to mention it in the description.